handler/effectiveschedule.go

```diff
@@ -46,69 +46,58 @@ func (s *ScheduleHandler) synchronizeEffectiveSchedulesResources() error {
+// It updates the context with either the existing schedule, or generates a new random schedule.
+func (s *ScheduleHandler) getOrGenerateEffectiveSchedule(ctx *deduplicationContext) {
+	existingSchedule, found := s.findExistingSchedule(ctx)
+	if found {
+		ctx.effectiveSchedule = existingSchedule
+		return
 	}
-	s.setEffectiveSchedule(jobType, randomizedSchedule, originalSchedule)
-	return randomizedSchedule
+	generatedSchedule, _ := s.createRandomSchedule(ctx.jobType, ctx.originalSchedule)
+	ctx.effectiveSchedule = generatedSchedule
 }
 
-// findExistingSchedule searches in the pre-fetched EffectiveSchedules and tries to find a generated schedule definition for the given schedule object and jobType.
+// findExistingSchedule searches in the cached EffectiveSchedules and tries to find a generated schedule definition for the given schedule object and jobType.
 // It returns empty string and false if none were found.
-func (s *ScheduleHandler) findExistingSchedule(jobType k8upv1alpha1.JobType, originalSchedule k8upv1alpha1.ScheduleDefinition) (k8upv1alpha1.ScheduleDefinition, bool) {
-	es, found := s.effectiveSchedules[jobType]
+func (s *ScheduleHandler) findExistingSchedule(ctx *deduplicationContext) (k8upv1alpha1.ScheduleDefinition, bool) {
+	es, found := s.effectiveSchedules[ctx.jobType]
 	if found {
 		for _, ref := range es.Spec.ScheduleRefs {
-			if s.schedule.IsReferencedBy(ref) && es.Spec.OriginalSchedule == originalSchedule {
+			if s.schedule.IsReferencedBy(ref) && es.Spec.OriginalSchedule == ctx.originalSchedule {
 				s.Log.V(1).Info("using generated schedule",
-					"name", k8upv1alpha1.GetNamespacedName(&es),
+					"name", k8upv1alpha1.GetNamespacedName(&es).String(),
 					"schedule", es.Spec.GeneratedSchedule,
-					"type", jobType)
+					"type", ctx.jobType)
 				return es.Spec.GeneratedSchedule, true
 			}
 		}
 	}
 	return "", false
 }
 
-// setEffectiveSchedule will create or update the EffectiveSchedule for the given jobType with the given schedule definition.
+// upsertEffectiveScheduleInternally will create or update the EffectiveSchedule for the given jobType with the given schedule definition.
 // The EffectiveSchedules aren't persisted or updated in this function, use synchronizeEffectiveSchedulesResources() for that.
-func (s *ScheduleHandler) setEffectiveSchedule(jobType k8upv1alpha1.JobType, schedule, originalSchedule k8upv1alpha1.ScheduleDefinition) {
-	es, found := s.effectiveSchedules[jobType]
+func (s *ScheduleHandler) upsertEffectiveScheduleInternally(ctx *deduplicationContext) {
+	es, found := s.effectiveSchedules[ctx.jobType]
 	if !found {
-		es = s.newEffectiveSchedule(jobType)
+		es = s.newEffectiveSchedule(ctx.jobType)
 	}
-	es.Spec.GeneratedSchedule = schedule
-	es.Spec.JobType = jobType
-	es.Spec.OriginalSchedule = originalSchedule
+	es.Spec.GeneratedSchedule = ctx.effectiveSchedule
+	es.Spec.JobType = ctx.jobType
```

Comment:
Not sure if it actually makes sense to update `JobType` and `backendString` here? Because by our definition this is what defines the scope of the deduplication. Instead of changing the existing one it should create a new one that would handle the new type and backend. Ig there are no more schedules with this type and backend combo it can be deleted.

Reply:
This piece of code is used for both creating and updating the ES ("upsert" being an abbrev for 'update or insert').
Upon updating, the context should be equal to what is already existing, so those lines become basically no-ops.

Reply:
my point is, the detection whether a backend or schedule has changed, should happen before this func is called.

Reply:
Fair enough. So with the implementation of backend detection it should also be a no-op?

Reply:
I think so, but I'd need to look into it.